### PR TITLE
BREAKING CHANGE: support <section>less editing

### LINKED ISSUE #3754: Don't require sections - let ReSpec handle them
**Is your feature request related to a problem? Please describe.**
Apparently, wrapping thing in `<section>` elements is a feature some folks don't like about ReSpec.

ReSpec has been opinionated about `<section>` from the very beginning, but the Markdown module automatically wrap headings into `<sections>`. 

**Describe the solution you'd like**

I wonder if we could take the `Builder` and the `structure()` method from Markdown.js and apply it to documents. 

This might actually already work by simply enabling: `format: "markdown"`, but it might be nice as a more general opt in, without running everything through the markdown parser.

### PR DESCRIPTION
**Pull request by marcoscaceres**
Closes #3754

Allows editing documents without requiring `<section>`.

This PR is mostly just a refactor. It moves "Builder" to its own module "sections"...

### CODE REVIEW

File: src/core/sections.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    section.remove();
+    elem.append(...section.childNodes);
+  } else {
+    elem.textContent = "";
+  }
+  elem.appendChild(structuredInternals);
+}
+
+export function run() {
+  restructure(document.body);
```
Comment (marcoscaceres) [inline src/core/sections.js:117]: Should we make it configurable? Like, run if "markdown" and "????".... or, just make it be the standard operating mode?

File: src/core/markdown.js [hunk 2]
```diff
@@ -182,118 +176,6 @@ function workaroundBlockLevelMarkdown(element, selector) {
...
-
-function structure(fragment, doc) {
-  function process(root) {
-    const stack = new Builder(doc);
-    while (root.firstChild) {
-      const node = root.firstChild;
-      if (node.nodeType !== Node.ELEMENT_NODE) {
-        root.removeChild(node);
-        continue;
-      }
```
Comment (marcoscaceres) [inline src/core/markdown.js:256]: This is the only thing I had to remove... it was discarding too much stuff (well, stuff that was causing tests to fail that were expecting comment nodes... but I think it also trashed TEXT nodes too).

File: tests/spec/w3c/linter-rules/required-sections-spec.js [hunk 3]
```diff
@@ -11,7 +11,11 @@ import { noTrackStatus } from "../../../../src/w3c/headers.js";
 import { requiresSomeSectionStatus } from "../../../../src/w3c/linter-rules/required-sections.js";
 
 describe("w3c — required-sections", () => {
+  beforeAll(() => {
+    jasmine.DEFAULT_TIMEOUT_INTERVAL *= 2;
```
Comment (marcoscaceres) [inline tests/spec/w3c/linter-rules/required-sections-spec.js:15]: I think the w3c server was being slow tonight... I don't think we actually need this.

File: tests/spec/w3c/headers-spec.js [hunk 4]
```diff
@@ -1962,9 +1962,6 @@ describe("W3C — Headers", () => {
       <section id="first-sub-section">
         <h3>Testing</h3>
       </section>
-      <p id="p3">
```
Comment (marcoscaceres) [inline tests/spec/w3c/headers-spec.js:1965]: The structuring code doesn't like this kind of thing... it makes a new section or something... this kind of thing shouldn't happen anyway, so 🤷‍♂️

File: tests/spec/core/markdown-spec.js [hunk 5]
```diff
@@ -22,7 +22,7 @@ describe("Core - Markdown", () => {
     });
     const foo = doc.getElementById("foo");
     expect(foo).toBeTruthy();
-    expect(foo.textContent).toBe("1. Foo");
+    expect(foo.textContent.trim()).toBe("1. Foo");
```
Comment (marcoscaceres) [inline tests/spec/core/markdown-spec.js:25]: Now that we don't throw out text nodes, I think the markdown parser generates some extraneous whitespace after the headings. No big deal though.

File: src/core/sections.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
+/**
+ * Module core/sections
+ *
+ * Adds sections to the document, based on the heading structure.
```
Comment (sidvishnoi) [inline src/core/sections.js:4]:
```suggestion
 * Adds <section>s to the document, based on the heading structure.
```

File: src/core/sections.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+          stack.addElement(node);
+      }
+    }
+    return stack.root;
+  }
+  return process(fragment);
+}
+
+/**
+ * Re-structure DOM around elem whose markdown has been processed.
```
Comment (sidvishnoi) [inline src/core/sections.js:98]: Comment needs update. Unrelated to markdown now.

File: profiles/w3c.js [hunk 8]
```diff
@@ -32,11 +32,12 @@ const modules = [
   import("../src/core/link-to-dfn.js"),
   import("../src/core/xref.js"),
   import("../src/core/data-cite.js"),
-  import("../src/core/webidl-index.js"),
   import("../src/core/render-biblio.js"),
   import("../src/core/dfn-index.js"),
   import("../src/core/contrib.js"),
+  import("../src/core/sections.js"),
```
Comment (sidvishnoi) [inline profiles/w3c.js:38]: Should add this one to other profiles I think. It's a core breaking change.

File: src/core/markdown.js [hunk 9]
```diff
@@ -29,17 +29,11 @@
...
- *       <section>
- *         <h3>Subtitle</h3>
- *         <p>Here's some text.</p>
- *       </section>
- *       <section>
- *         <h3>Another subtitle</h3>
- *         <p>More text.</p>
- *       </section>
- *     </section>
+ *       <h3>Subtitle</h3>
```
Comment (sidvishnoi) [inline src/core/markdown.js:33]: Comment body above needs an update too.

File: src/core/sections.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        case "section":
+          stack.addSection(node, process);
+          break;
+        default:
+          stack.addElement(node);
+      }
+    }
+    return stack.root;
+  }
+  return process(fragment);
```
Comment (sidvishnoi) [inline src/core/sections.js:94]: Nit: we can un-nest this nested function I guess.

File: src/core/sections.js (hunk 1 above)
Comment (sidvishnoi) [reply src/core/sections.js:117]: I think making it default should be fine, as long as resultant output is same.

File: tests/spec/core/id-headers-spec.js [hunk 11]
```diff
@@ -33,7 +33,7 @@ describe("Core - ID headers", () => {
 
   it("sets an id on header", () => {
     const h2 = doc.querySelector("#t0 h2");
-    expect(h2.id).toBe("foo");
+    expect(h2.id).toBe("x1-foo");
```
Comment (sidvishnoi) [inline tests/spec/core/id-headers-spec.js:36]: Oh. This seems to be a breaking change. IIUC, it's supposed to be adding id `"foo"` to `<section>` now (but that section already has an id)?

File: tests/spec/core/id-headers-spec.js [hunk 12]
```diff
@@ -46,7 +46,7 @@ describe("Core - ID headers", () => {
       expect(intro.getAttribute("aria-label")).toBe("Permalink for Section");
       const custom = doc.querySelector("#custom-id a.self-link");
       expect(custom.getAttribute("aria-label")).toBe(
-        "Permalink for Section 2."
+        "Permalink for Section 3."
```
Comment (sidvishnoi) [inline tests/spec/core/id-headers-spec.js:49]: What happened here?

File: tests/spec/core/markdown-spec.js (hunk 5 above)
Comment (sidvishnoi) [reply tests/spec/core/markdown-spec.js:25]: Should investigate I think.

File: tests/spec/core/sections-spec.js [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+      <h5 id="h5">Section 5</h5>
+      <h6 id="h6">Section 6</h6>
+    `;
+    const ops = makeStandardOps(null, body);
+    const doc = await makeRSDoc(ops);
+    for (let i = 2; i <= 6; i++) {
+      const h = doc.getElementById(`h${i}`);
+      const section = h.parentElement;
+      expect(section.localName).toBe("section");
+      expect(section.id).toBe(`section-${i}`);
```
Comment (sidvishnoi) [inline tests/spec/core/sections-spec.js:22]: Nit: add `withContext`.

File: tests/spec/core/sections-spec.js [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    }
+  });
+
+  it("it handles when some headings are wrapped", async () => {
```
Comment (sidvishnoi) [inline tests/spec/core/sections-spec.js:26]:
```suggestion
  it("handles when some headings are wrapped", async () => {
```

File: tests/spec/core/sections-spec.js [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+      const section = h.parentElement;
+      expect(section.localName).toBe("section");
+      expect(section.id).toBe(`section-${i}`);
+    }
+  });
+
+  it("retains different nodes when wrapping", async () => {
+    const body = `
+      <h2 id="h2">Section 2</h2>
+        <p>paragraph 2</p><!--comment 2-->
```
Comment (sidvishnoi) [inline tests/spec/core/sections-spec.js:53]: Is this indent on `<p>` intentional/meaningful?

File: tests/spec/w3c/headers-spec.js (hunk 4 above)
Comment (sidvishnoi) [reply tests/spec/w3c/headers-spec.js:1965]: I think this should be wrapped into a section of its own?

Comment (sidvishnoi) [reply tests/spec/w3c/headers-spec.js:1965]: Section without heading that the linter shouts at.

File: tests/spec/w3c/linter-rules/required-sections-spec.js (hunk 3 above)
Comment (sidvishnoi) [reply tests/spec/w3c/linter-rules/required-sections-spec.js:15]: Yeah, let's not forget to remove this. Or, we should probably set it at config level as it's a "timeout".

File: tests/spec/core/id-headers-spec.js (hunk 11 above)
Comment (sidvishnoi) [reply tests/spec/core/id-headers-spec.js:36]: Concerned with any sort of xref/webref breakage mainly.

File: tests/spec/core/sections-spec.js (hunk 15 above)
Comment (marcoscaceres) [reply tests/spec/core/sections-spec.js:53]: Nope. Will fix.

File: tests/spec/core/markdown-spec.js (hunk 5 above)
Comment (marcoscaceres) [reply tests/spec/core/markdown-spec.js:25]: Yeah, we have some inconsistency on how the ids are being assigned to sections and headers somewhere... I'll investigate further.

That also explains `.trim()` in a few places above... the `<section>`s are now getting the ids that were on the heading elements.

File: tests/spec/w3c/headers-spec.js (hunk 4 above)
Comment (marcoscaceres) [reply tests/spec/w3c/headers-spec.js:1965]: I think that gets messy/confusing, tbh... plus it's too much of an edge case to worry about.

Comment (sidvishnoi) [reply tests/spec/w3c/headers-spec.js:1965]: Let's skip it then

File: tests/spec/core/id-headers-spec.js (hunk 11 above)
Comment (marcoscaceres) [reply tests/spec/core/id-headers-spec.js:36]: This is consistent with "normal" mode, so it should be fine. Also, both ids (the one on the section and the one on the header are retained, so nothing will break.)

File: tests/spec/core/id-headers-spec.js (hunk 12 above)
Comment (marcoscaceres) [reply tests/spec/core/id-headers-spec.js:49]: This is correct... removing the `<p>BLAH</p>` above (which we won't support), causes the sections to be ordered correctly... so this is now "3. Pass":

Screen Shot 2022-01-31 at 12 06 58 pm

**Comment by marcoscaceres** [comment]
Ok, I think we are good to go here.

**Review by sidvishnoi — APPROVED** [review]
LGTM. Not sure if feat or breaking change anymore though.

**Comment by marcoscaceres** [comment]
> LGTM. Not sure if feat or breaking change anymore though.

Same... I figured it's a breaking change because it breaks the profiles... so if someone out there has their own profile, this would break.
